exporters/trace/src/main/java/com/google/cloud/opentelemetry/trace/TraceExporter.java

```diff
@@ -15,134 +15,72 @@
+        return new TraceExporter(configBuilder);
+    }
 
-      return new TraceExporter(
-          projectId,
-          new CloudTraceClientImpl(TraceServiceClient.create(builder.build())),
-          configuration.getAttributeMapping(),
-          configuration.getFixedAttributes());
+    private SpanExporter createActualTraceExporter() throws IOException {
+        return InternalTraceExporter.createWithConfiguration(this.customTraceConfigurationBuilder.build());
     }
-    return TraceExporter.createWithClient(
-        projectId,
-        new CloudTraceClientImpl(TraceServiceClient.create(stub)),
-        configuration.getAttributeMapping(),
-        configuration.getFixedAttributes());
-  }
 
-  private static TraceExporter createWithClient(
-      String projectId,
-      CloudTraceClient cloudTraceClient,
-      ImmutableMap<String, String> attributeMappings,
-      Map<String, AttributeValue> fixedAttributes) {
-    return new TraceExporter(projectId, cloudTraceClient, attributeMappings, fixedAttributes);
-  }
+    public static SpanExporter createWithDefaultConfiguration() {
+        return generateStubTraceExporter(TraceConfiguration.builder());
+    }
 
-  TraceExporter(
-      String projectId,
-      CloudTraceClient cloudTraceClient,
-      ImmutableMap<String, String> attributeMappings,
-      Map<String, AttributeValue> fixedAttributes) {
-    this.projectId = projectId;
-    this.cloudTraceClient = cloudTraceClient;
-    this.projectName = ProjectName.of(projectId);
-    this.translator = new TraceTranslator(attributeMappings, fixedAttributes);
-  }
+    public static SpanExporter createWithConfiguration(TraceConfiguration.Builder configBuilder) {
+        return generateStubTraceExporter(configBuilder);
+    }
 
-  @Override
-  public CompletableResultCode flush() {
-    // We do no exporter buffering of spans, so we're always flushed.
-    return CompletableResultCode.ofSuccess();
-  }
+    @Deprecated
+    public static SpanExporter createWithConfiguration(TraceConfiguration configuration) throws IOException {
+        return InternalTraceExporter.createWithConfiguration(configuration);
+    }
 
-  @Override
-  public CompletableResultCode export(Collection<SpanData> spanDataList) {
-    List<Span> spans = new ArrayList<>(spanDataList.size());
-    for (SpanData spanData : spanDataList) {
-      spans.add(translator.generateSpan(spanData, projectId));
+    @Override
+    public CompletableResultCode flush() {
+        if (internalTraceExporter == null) {
+            return CompletableResultCode.ofFailure();
+        }
+        return internalTraceExporter.flush();
     }
 
-    cloudTraceClient.batchWriteSpans(projectName, spans);
-    return CompletableResultCode.ofSuccess();
-  }
+    @Override
+    public CompletableResultCode export(@Nonnull Collection<SpanData> spanDataList) {
+        try {
+            if (internalTraceExporter == null) {
```

Comment:
FYI - this type of delegation is not threadsafe. 


I suggest you use an `AtomicReference<SpanExporter>`, and you can do something like:

```
while (internalTraceExporter.get() == null) {
  internalTraceExporter.compareAndSet(null, createActualTraceExporter());
}
```

This will ensure that trace exporter is created (Possibly more than once, but only one is kept), and the AtomicReference ensures you'll be threadsafe.

You can also look up "typesafe singleton pattern in Java" for other options.

Reply:
Yup, that's a good point. I will make the change.